### DISCUSSION TITLE
How to dynamically set DataGrid row color based on one of its entries?

### DISCUSSION THREAD

**Discussion by Mersid** [AvaloniaUI/Avalonia#9875] [Q&A] [answered]
Heya! I'm a little new to this whole business. I've spent the last two days banging my head against this problem, to little avail. Hopefully you guys can help a noob out here!

The question is how do I set a DataGrid row's background color based on the value of its entries? My UI looks like this, currently
![image](https://user-images.githubusercontent.com/30786211/210687938-b70606e3-408e-4ac4-828c-72ec9ef5b3fd.png)

These entries get populated when the Scan button on the bottom right is clicked. Upon clicking, the values will be populated, with Files, Folders, and Total set to 0, and Status set to "Scanning". A background task runs these and will set the Status to "Scanned" when it is done. This is handled via binding. What I want to have happen is that when the Status for a specific row is "Scanned" (or one of any other statuses), the entire row gets colored for easy identification. How would I do this? I've tried various solutions, and none seem to work.

1. Use events on the DataGrid, such as LoadingRow, and handle it in code: This sets the color, but it doesn't work as the event doesn't get called when the Status for an entry is updated, so it retains the old color.
2. Binding RowBackground and AlternatingRowBackground to a property that I generate from the model: I've tried messing around with this. Often it straight up doesn't work, in the iterations where it does the color doesn't update when changes are applied (same as in 1).
3. DataGrid.Styles and UserControl.Styles with a Converter: Again, it only works for the initial run when the rows are created. When updated, the color does not update.

In most of these scenarios, I noticed that clicking any of the headers to sort by that column will update all the colors, and manipulating the UI's dimensions so that some rows falls outside the screen updates those rows (because they get recreated), but obviously I want them to update as soon as the associated model updates. How would I do that?

Using technique 3, I got it to something like this:
![image](https://user-images.githubusercontent.com/30786211/210689337-b07c5f88-05aa-4dba-a61b-10bbce141e6a.png)

The scanned entries are highlighted with a different color, except for one, because they only update when I clicked a row header to sort them.

Here's the relevant code files, if it helps!
The MainView.axaml file, which describes the view shown in the above screenshots
```xml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:viewModels="clr-namespace:CacheCleaner.ViewModels"
             xmlns:local="clr-namespace:CacheCleaner"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="CacheCleaner.Views.MainView">
    
    <Design.DataContext>
        <viewModels:MainViewViewModel/>
    </Design.DataContext>
    
    <UserControl.Resources>
        <local:RowConverter x:Key="RowConverter" />
    </UserControl.Resources>
    
    <UserControl.Styles>
        <Style Selector="DataGridRow">
            <Setter Property="Background" Value="{Binding  Converter={StaticResource RowConverter}}" />
        </Style>
    </UserControl.Styles>
    
    <Grid ShowGridLines="False" ColumnDefinitions="*, *, *, *" RowDefinitions="*, *, *" Margin="0">
        <DataGrid Grid.Row="0" Grid.Column="0" Grid.RowSpan="2" Grid.ColumnSpan="4"
                  AutoGenerateColumns="false" CanUserReorderColumns="True" CanUserResizeColumns="True" CanUserSortColumns="True" IsReadOnly="True"
                  DataContext="{Binding Items}"
                  Items="{Binding}">
            <DataGrid.Columns>
                <DataGridTextColumn Header="Path" Binding="{Binding Path}"/>
                <DataGridTextColumn Header="Files" Binding="{Binding Files}" />
                <DataGridTextColumn Header="Folders" Binding="{Binding Folders}" />
                <DataGridTextColumn Header="Total" Binding="{Binding Total}" />
                <DataGridTextColumn Header="Status" Binding="{Binding Status}" />
            </DataGrid.Columns>
        </DataGrid>
        
        <StackPanel Grid.Row="2" Grid.Column="3" VerticalAlignment="Bottom">
            <Button Command="{Binding Scan}" IsEnabled="{Binding ScanButtonEnabled}" VerticalAlignment="Stretch" HorizontalAlignment="Stretch">
                <Label VerticalAlignment="Center" HorizontalAlignment="Center">Scan</Label>
            </Button>
            <Button Command="{Binding Delete}" IsEnabled="{Binding DeleteButtonEnabled}" VerticalAlignment="Stretch" HorizontalAlignment="Stretch">
                <Label VerticalAlignment="Center" HorizontalAlignment="Center">Delete</Label>
            </Button>
        </StackPanel>
            
    </Grid>
</UserControl>

```

The associated ViewModel:
```cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Avalonia.Logging;
using Avalonia.Media;
using Avalonia.Threading;
using CacheCleaner.Models;
using ReactiveUI;

namespace CacheCleaner.ViewModels
{
	public class MainViewViewModel : ViewModelBase
	{
		public CleaningEntryModel[] CleaningEntryModelTest => new[]
		{
			new CleaningEntryModel() {Path = Environment.ExpandEnvironmentVariables("%temp%"), Files = 123, Folders = 456, Status = "Not Scanned"}
		};
		
		public CleaningEntry[] CleaningEntries => JsonSerializer.Deserialize<CleaningEntry[]>(File.ReadAllText("Resources/CleaningEntries.json"), new JsonSerializerOptions {PropertyNameCaseInsensitive = true})!;

		public ObservableCollection<CleaningEntryModel> Items { get; } = new ObservableCollection<CleaningEntryModel>();
		
		// Is not a list because it is updated across threads
		private readonly ConcurrentBag<CleaningEntryFileEnumeration> fileEnumerations = new ConcurrentBag<CleaningEntryFileEnumeration>();

		private readonly ConcurrentBag<CleaningEntryExecutionResult> executionResults =
			new ConcurrentBag<CleaningEntryExecutionResult>();

		private bool scanButtonEnabled = true;
		public bool ScanButtonEnabled
		{
			get => scanButtonEnabled;
			set => this.RaiseAndSetIfChanged(ref scanButtonEnabled, value);
		}

		private bool deleteButtonEnabled;

		public bool DeleteButtonEnabled
		{
			get => deleteButtonEnabled;
			set => this.RaiseAndSetIfChanged(ref deleteButtonEnabled, value);
		}
		
		public void Scan()
		{
			ScanButtonEnabled = false;
			DeleteButtonEnabled = false;
			List<Task> tasks = new List<Task>();
			Items.Clear();
			fileEnumerations.Clear();
			foreach (CleaningEntry cleaningEntry in CleaningEntries)
			{
				CleaningEntryModel entryModel = new CleaningEntryModel
				{
					Path = FileDeleter.ExpandNormalize(cleaningEntry.Directory),
					Status = "Scanning"
				};
				Items.Add(entryModel);


				tasks.Add(new Task(() =>
				{
					CleaningEntryFileEnumeration enumeration = FileDeleter.GetEntryFileEnumeration(cleaningEntry).Result;
					entryModel.Files = enumeration.Files.Count;
					entryModel.Folders = enumeration.Folders.Count;
					entryModel.Status = "Scanned";
					fileEnumerations.Add(enumeration);
				}));
			}

			new Task(() =>
			{
				Task.WaitAll(tasks.ToArray());
				ScanButtonEnabled = true;
				DeleteButtonEnabled = true;
			}).Start();
			
			tasks.ForEach(t => t.Start());
		}

		public void Delete()
		{
			DeleteButtonEnabled = false;
			List<Task> tasks = new List<Task>();
			foreach (CleaningEntryFileEnumeration entry in fileEnumerations)
			{
				CleaningEntryModel uiModelEntry = Items.First(t => t.Path == entry.Path);
				uiModelEntry.Status = "Deleting";
				tasks.Add(new Task(() =>
				{
					CleaningEntryExecutionResult entryExecutionResult = FileDeleter.ExecuteCleaningEntry(entry);

					uiModelEntry.Status =
						$"Deleted {entryExecutionResult.SuccessFiles.Count} files and {entryExecutionResult.SuccessFolders.Count} " +
						$"folders. Failed to delete {entryExecutionResult.FailureFiles.Count} files and " +
						$"{entryExecutionResult.FailureFolders.Count} folders";
					
					executionResults.Add(entryExecutionResult);
				}));
			}
			
			tasks.ForEach(t => t.Start());
			
		}
	}
}
```

The model that backs the DataGrid entries (ignore the Brush property - it was for another test that didn't work):
```cs
using System.ComponentModel;
using Avalonia.Media;

namespace CacheCleaner.Models;

public class CleaningEntryModel : INotifyPropertyChanged
{
	public string Path { get; init; }
	
	private int files;
	public int Files
	{
		get => files;
		set
		{
			files = value;
			PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Files)));
			PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Total)));
		}
	}

	private int folders;
	public int Folders
	{
		get => folders;
		set
		{
			folders = value;
			PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Folders)));
			PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Total)));

		}
	}
	public int Total => Files + Folders;
	private string status;
	public string Status
	{
		get => status;
		set
		{
			status = value;
			PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Status)));
			PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Brush)));
		}
	}

	// public ISolidColorBrush Brush => Status == "Scanned" ? Brushes.Wheat : Brushes.Aquamarine;
	public ISolidColorBrush Brush => Status == "Scanned" ? Brushes.Wheat : Brushes.Transparent;

	public event PropertyChangedEventHandler? PropertyChanged;
}
```

And the converter
```cs
using System;
using System.Globalization;
using Avalonia.Data.Converters;
using Avalonia.Media;
using CacheCleaner.Models;

namespace CacheCleaner;

public class RowConverter : IValueConverter
{
	public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
	{
		if (value is CleaningEntryModel { Status: "Scanned" })
		{
			return Brushes.Wheat;
		}
		
		return Brushes.Transparent;
	}

	public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
	{
		return null;
	}
}
```

Hopefully this is enough information to help resolve the problem. I appreciate the support!

**Comment by maxkatz6** [comment]
> DataGrid.Styles and UserControl.Styles with a Converter: Again, it only works for the initial run when the rows are created. When updated, the color does not update.

That's a possible way.
Your problem in the code that you pass whole row data model to the binding and converter. As expected, resulting value is never changed, as data model object by itself is not changed. Instead, you need to pass model property in the binding. Something like `{Binding Status, Converter={...}}`. And read status in the converter directly from the value. If you need to pass multiple values, use MultiBinding.

> Use events on the DataGrid, such as LoadingRow, and handle it in code: This sets the color, but it doesn't work as the event doesn't get called when the Status for an entry is updated, so it retains the old color.

That's another way that is also possible. You just need to create a binding in the LoadingRow handler.
```csharp
        public void RowsDataGrid_LoadingRow(object? sender, DataGridRowEventArgs e)
        {
            var row = e.Row;
            row.Bind(DataGridRow.BackgroundProperty, new Binding("Status", BindingMode.OneWay)
            {
                   Converter = ...
            });
       }
```
I didn't test this code, but have used similar in the past.


> Binding RowBackground and AlternatingRowBackground to a property that I generate from the model: I've tried messing around with this. Often it straight up doesn't work, in the iterations where it does the color doesn't update when changes are applied (same as in 1).

Not sure how it is possible, if you need to bind row data to a row control, while RowBackground is a property on the parent DataGrid with its own parent data context.

**Comment by Mersid** [comment]
Thanks, Max! Unfortunately, it appears that I may be doing something wrong, or the proposed solutions do not work.
I have made the modifications you mentioned:  The binding now specifies the Status property, and the converter has been updated to work with it:
```cs
	public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
	{
		string? status = value as string;
		
		if (status is "Scanned")
		{
			return Brushes.Wheat;
		}
		
		return Brushes.Transparent;
	}
```

It unfortunately did not work. The issue remains the same: the color change only applies when the rows are regenerated (by sorting, moving off-screen, etc).

I also tried the events approach:
```cs

	private void DataGrid_OnLoadingRow(object? sender, DataGridRowEventArgs e)
	{
		DataGridRow row = e.Row;
		row.Bind(DataGridRow.BackgroundProperty, new Binding("Status", BindingMode.OneWay)
		{
			Converter = new RowConverter()
		});
	}
```

And while the debugger does show the points being hit, the same issue persists: Updates are only applied when the rows are regenerated.

Do you know why this could be the case?

Again, thanks for your willingness to help - I've been trying to fix this problem for a while now.

**Comment by Mersid** [comment, marked as answer]
I suppose another question worth asking is whether there is a way to force re-rendering of the DataGridRows. I know that clicking any of the headers to sort induces this; is there a way to cause this programatically without sorting anything? Such a method should theoretically update all the backgrounds.

**Reply by timunie** [reply]
I still think a style for DataGridRow would be the easier way. Can you put together a simple sample and upload it to a new GithubRepo?

**Reply by Mersid** [reply]
Thanks for being willing to take a look, Tim! I'm surprised at the responsiveness of this community :)

I've uploaded my project to Github (the whole thing). I've made a change to it so it applies a style for DataGridRow to set its background, as you've suggested. If it exhibits the same behaviour as it does for me, then when you click the "Scan" button, some directories will show up as scanning, then proceed to scanned once it finishes traversing those directories. Notice that the color stays the default, unless you use a sort button or shrink the entries offscreen. Then any entry with a status of scanned will show up in a wheat color. What I'm aiming for is the color to transition automatically as the status changes; kind of like conditional formatting in Excel, if you will.

https://github.com/Mersid/CacheCleaner

Again. much appreciated!

**Reply by timunie** [reply]
Please check https://github.com/Mersid/CacheCleaner/pull/1 

Things I changed:

Use CompiledBindings where possible (optional, but helps in finding issues), see https://docs.avaloniaui.net/docs/data-binding/compiledbindings
Fix a wrong string comparison. Use == and not is operator, see https://learn.microsoft.com/en-US/dotnet/csharp/language-reference/operators/is
Change Style-Selector to use DataGridCell. Found in DevTools ([F12] in debug mode) that the background is acutally applied to DataGridCell works better.

Happy coding
Tim

**Reply by Mersid** [reply]
God damn! You actually did it! Thanks so much for your assistance! It's late here, so I'll take a look tomorrow to see how it's done and what I can learn from it. I do notice from a quick glance, however, some stuff that I didn't notice from the Avalonia documentation. Was it simply missing, or did I simply lack the requisite knowledge to notice it? For example, all the x:DataType=... and stuff seems to have also fixed some issues with paths not being found by the IDE. Is there documentation on this?
Much love <3

**Reply by timunie** [reply]
Can you please tell me which information exactly was not discoverable for you? Maybe we can imporve the docs together.

**Reply by Mersid** [reply]
Heya Tim! Again, thanks for your help. I do have another question, though. When setting the rows to a color, I noticed that it's no longer possible to determine which rows are selected by a user - normally they turn a shade of blue, but with colors applied it no longer shows up. I've tried appkying another style selector with the :selected and :pointerover pseudoclasses, but when selecting DataGridRow, it simply doesn't work, and when selecting DataGridCell, it only applies the color to the specific cell aimed/clicked on, not the whole row. How would I solve that? 
I added this, basically.
```xml
<Style Selector="DataGridCell:selected" x:DataType="models:CleaningEntryModel">
                    <Setter Property="Background" Value="Aquamarine" />
</Style>
```
I'd want to either set entire rows' selected entries to a different color, or set it to transparent again and the natural hover-over color should appear. However, I want to color the entire row, not just the cell.

As for information, it appears that technically almost all the relevant information is available on the documentation site, but if, for example, you did not give me the relevant snippets to have fixed it, I already looked through those documentation sections and skimmed right over it. It's quite subjective, but perhaps an issue is simply the docs lack sufficient relevant examples or show how to do operations when it comes to the less common issues. I'm no expert at writing documentation so it's hard to express in words. I also did not find any information on DataGridRows and DataGridCells in the DataGrid section entry; I suspect I am not the first, nor the last, person to run against why, for example, applying a style to a row does not take immediate effect - after all, changing a model will reflect its changes in the UI immediately.

**Reply by Mersid** [reply]
I think I managed to remedy the issue. I simply set a custom brush with partial transparency. It turns out the color format was in ARGB, and not RBGA as I tested earlier. May still be worth learning how to solve it with style selectors and pseudoclasses, though :)

**Reply by timunie** [reply]
Yes, samples are missing. I agree. That's why we started to create a sample Repo linked to relevant doc pages. But it will take some time to cover almost everything. 

Meanwhile you are welcome to ask here, on telegram or gitter.

**Reply by FraserElectronics** [reply]
Can I ask a follow-up on this?

I have used the same principal as the changes you made to CacheCleaner in that I have a boolean field bound to my data grid and use a converter to return a colour based on whether the value is true or false.  I use the data grid style selector as per your example and it *almost* works in that the row changes colour but only if I scroll the grid or change to a different view and back.
My boolean flag is a field of a class that derives from ObservableObject and the field itself is marked as an ObservableProperty. I assumed that because it is an observable property, any change would be immediately reflected in the grid but not so !

What am I missing?

Andy

**Reply by jp2masa** [reply]
Are you setting the field or the generated property? Setting the field won't notify the UI.

**Reply by FraserElectronics** [reply]
@jp2masa Yes, I am setting the generated property, not the backing field.

Andy

**Comment by shukriadams** [comment]
Would be great if there was some sort of write-up of the solution, sitting with the same problem myself and it's hard to tease out the salient facts from this thread : how does one, using axaml, set the row color of a individual datagrid row, based on a property of the data object associated with that row?

**Reply by timunie** [reply]
ControlCatalog in src/samples has colord cells based on values.

**Reply by shukriadams** [reply]
I want to set row color, not cell color, but maybe Avalonia doesn't expose row style? I found https://github.com/AvaloniaUI/Avalonia/blob/dac4ae422da17afb717c8e60e0728a1d8b1b4751/samples/ControlCatalog/Pages/DataGridPage.xaml#L93 but that's not what I'm looking for. I must confess, I have no idea what to even start looking for, as I have no reference for how databinding and style are combined in Avalonia.

Maybe I should phrase the question differently. 

			<DataGrid
				ItemsSource="{Binding Packages}">
				<DataGrid.Columns>
						<DataGridTextColumn Binding="{Binding Package.Id}"/>
						<DataGridTextColumn Binding="{Binding Package.State}"/>
						<DataGridTemplateColumn Header="Tags" Width="300">
                                                     ... template column definitions here ...
                                                </DataGridTemplateColumn>
					</DataGrid.Columns>
			</DataGrid>

I have a Datagrid above, which displays data for a list of objects, across different types of cells, and the data are constantly changing. I want to dynamically change the color for an entire row in the grid, based on the value of the data property "State" (in the 2nd column). State changes constantly, and row color must update with it. How would I do that? 

In an ideal world, for someone with an HTML mindset, there would be some kind static style like `DataGridRow.failingState` with a setter like `Background:Red`. Then from code we could append `failingState` to the `Classes` property on some DataGridRow, and the row would turn red. When we removed failingState from Classes on that DataGridRow, the row would revert to its default color. Is something like this possible? It appears we cannot programmatically access DataGridRows by looping over a collection of them, even though vanilla WPF allows this.

**Reply by timunie** [reply]
Background is set by Package.State and a Converter.

**Reply by shukriadams** [reply]
Ok, I got it working, and it's still a bit confusing, but I hope this helps someone else out.

You need to bind the row load event
 
    <DataGrid LoadingRow="OnLoadRow" />

Then in codebehind bind to both the row and datacontext objects
 
        private void OnLoadRow(object sender, DataGridRowEventArgs e) 
        {
            DataGridRow row = e.Row;
            MyDataType myDataObject = e.Row.DataContext as MyDataType;
            row.Bind(DataGridRow.BackgroundProperty, 
                new Binding("MyDataProperty", BindingMode.OneWay) { Converter = new RowStatusConverter() }, 
                myDataObject);
        }

And then finally you define a RowStatusConverter converter class that implements IValueConverter, which will receive the value of MyDataType.MyDataProperty each time that property changes, assuming you've made it Observable etc etc.

This was actually simple enough to figure out from the thread above, but only once one accepts that the converter can't be bound in markup (which is the WPF convention, or am I mistaken?). Plus, the only way we can get at the ever-elusive DataGridRow is via the Row Load handler, which also seems to break from WPF standard. I'm sure that ship has already sailed, but not being able to get at a Row Items collection on the datagrid is a bit of a drag.

Also, once I knew how to do this, I could search the code examples in this repo with a bit more focus, and found no mention of the above approach, nor anything that combines Style + Binding of a list in a repeater. This approach of binding to just a single property, in this case Background, is workable, but it means we can change only one style property per binding. My earlier question still stands : can we somehow attach and detach a class string name to a row element with a binder, and if so, exactly how do we that (with an actual example).

Better and simpler examples are nice, but good documentation is the best. Sorry if I seem pedantic, this just took way longer than it should have. Thanks.

**Reply by timunie** [reply]
Can you file a minimal sample App here? The idea is the use a Style for Row in XAML. Converter works but you need to set x:DataType for compiled bindings
